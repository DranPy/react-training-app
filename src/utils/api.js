import axios from 'axios';
// NOTE: add a symlink
// import { host, port } from '../json-server/db.json';

//const URL = `http://${host}:${port}`;
const baseURL = `http://localhost:4000`;

export default axios.create({ baseURL });
