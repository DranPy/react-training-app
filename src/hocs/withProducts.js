import React, { Component } from 'react';

import api from 'utils/api-helper';

const withProducts = options => WrappedComponent => {
  return class extends Component {
    static displayName = `withProducts(${WrappedComponent.displayName || WrappedComponent.name})`;

    state = {
      isLoading: false,
      products: [],
    };

    fetchProducts = async () => {
      this.setState({ isLoading: true });
      const products = await api.getAllProducts();
      this.setState({
        isLoading: false,
        products,
      });

      return products;
    };

    componentDidMount() {
      if (options && options.isFetchImmediately) {
        this.fetchProducts();
      }
    }

    render() {
      const { isLoading, products } = this.state;

      const newProps = {
        fetchProducts: this.fetchProducts,
        isLoading,
        products,
      };

      return <WrappedComponent {...this.props} {...newProps} />;
    }
  };
};

export default withProducts;
