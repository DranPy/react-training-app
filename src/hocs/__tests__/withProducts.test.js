import { shallow, mount } from 'enzyme';
import React, { Component } from 'react';

import withProducts from '../withProducts';
import api from '../../utils/api-helper';

jest.mock('../../utils/api-helper');

describe('HOC withProducts', () => {
  let WrapperComponent;

  class MockComponent extends Component {
    render() {
      return <div>Component</div>;
    }
  }

  it('should correct render', () => {
    WrapperComponent = withProducts()(MockComponent);
    const wrapper = shallow(<WrapperComponent />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch products immediately', () => {
    WrapperComponent = withProducts({ isFetchImmediately: true })(MockComponent);
    const wrapper = shallow(<WrapperComponent />, { disableLifecycleMethods: true });
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchProducts');
    instance.componentDidMount();
    expect(instance.fetchProducts).toHaveBeenCalledTimes(1);
  });

  it('should not fetch products immediately', () => {
    WrapperComponent = withProducts({ isFetchImmediately: false })(MockComponent);
    const wrapper = shallow(<WrapperComponent />, { disableLifecycleMethods: true });
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchProducts');
    instance.componentDidMount();
    expect(instance.fetchProducts).not.toHaveBeenCalledTimes(1);
  });

  it('should fetch products to state', async () => {
    WrapperComponent = withProducts()(MockComponent);
    const mockProducts = [{ name: 'test 1' }, { name: 'test 2' }];
    api.getAllProducts.mockReturnValue(Promise.resolve(mockProducts));
    const wrapper = mount(<WrapperComponent />);
    const instance = wrapper.instance();
    await instance.fetchProducts();
    expect(wrapper.state('products')).toEqual(mockProducts);
  });
});
