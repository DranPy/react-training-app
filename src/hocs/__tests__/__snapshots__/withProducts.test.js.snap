// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC withProducts should correct render 1`] = `
<MockComponent
  fetchProducts={[Function]}
  isLoading={false}
  products={Array []}
/>
`;
